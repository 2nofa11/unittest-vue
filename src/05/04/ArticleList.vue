<script setup lang="ts">
import ArticleListItem, { type ItemProps } from "./ArticleListItem.vue";

type Props = {
  items: ItemProps[];
};

defineProps<Props>();
</script>

<template>
  <div>
    <h2 data-test="article-list-heading">記事一覧</h2>
    <ul v-if="items.length" data-test="article-list">
      <ArticleListItem
        v-for="item in items"
        :key="item.id"
        :id="item.id"
        :title="item.title"
        :body="item.body"
      />
    </ul>
    <p v-else data-test="no-articles-message">投稿記事がありません</p>
  </div>
</template>
