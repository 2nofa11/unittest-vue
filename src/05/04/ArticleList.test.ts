import { describe, expect, test } from "@jest/globals";
import { mount } from "@vue/test-utils";
import ArticleList from "./ArticleList.vue";
import { items } from "./fixture";

describe("ArticleList.vue", () => {
  test("タイトルの表示", () => {
    const wrapper = mount(ArticleList, { props: { items } });
    const heading = wrapper.find('[data-test="article-list-heading"]');
    expect(heading.exists()).toBe(true);
    expect(heading.text()).toBe("記事一覧");
  });

  test("items の数だけ一覧表示される", () => {
    const wrapper = mount(ArticleList, { props: { items } });
    expect(wrapper.findAll('[data-test="article-list-item"]').length).toBe(
      items.length
    );
  });

  test("一覧アイテムが空のとき「投稿記事がありません」が表示される", () => {
    const wrapper = mount(ArticleList, { props: { items: [] } });
    const list = wrapper.find('[data-test="article-list"]');
    expect(list.exists()).toBe(false);
    const message = wrapper.find('[data-test="no-articles-message"]');
    expect(message.exists()).toBe(true);
    expect(message.text()).toContain("投稿記事がありません");
  });

  test("Snapshot: items の数だけ一覧表示される", () => {
    const wrapper = mount(ArticleList, { props: { items } });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
