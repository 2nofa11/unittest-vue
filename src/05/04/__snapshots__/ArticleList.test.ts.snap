// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleList.vue Snapshot: items の数だけ一覧表示される 1`] = `
"<div>
  <h2 data-test="article-list-heading">記事一覧</h2>
  <ul data-test="article-list">
    <li data-test="article-list-item">
      <h3 data-test="article-title">TypeScript を使ったテストの書き方</h3>
      <p data-test="article-body">テストを書く時、TypeScript を使うことで、テストの保守性が向上します…</p><a href="/articles/howto-testing-with-typescript" data-test="article-link">もっと見る</a>
    </li>
    <li data-test="article-list-item">
      <h3 data-test="article-title">Next.js の Link コンポーネント</h3>
      <p data-test="article-body">Next.js の画面遷移には、Link コンポーネントを使用します…</p><a href="/articles/nextjs-link-component" data-test="article-link">もっと見る</a>
    </li>
    <li data-test="article-list-item">
      <h3 data-test="article-title">Jest ではじめる React のコンポーネントテスト</h3>
      <p data-test="article-body">Jest は単体テストとして、UIコンポーネントのテストが可能です…</p><a href="/articles/react-component-testing-with-jest" data-test="article-link">もっと見る</a>
    </li>
  </ul>
</div>"
`;
